File: packages/sdk-solana/src/math/lmsr.ts
// TypeScript port of `sooth_core`'s `math/{lmsr.rs, wad.rs}`. WAD = 1e18.
//
// It exists so a quote can be computed without a round trip, which means it
// must agree with the on-chain result exactly — a divergence is not a rounding
// difference, it is a trade that quotes one price and settles at another.
//
// Invariants carried over from the Rust:
//
//   - WAD multiplication and division match the Rust 256-bit-intermediate
//     semantics. JS bigint handles arbitrary precision natively, so the
//     limb-based math is collapsed to direct bigint ops; the rounding is
//     toward zero (same as the Rust port's `i128` truncating division).
//
//   - `expWad` saturates the negative tail to 0 past `EXP_MAX_INPUT_WAD`
//     (= 64·WAD). This is what makes the log-sum-exp shifted `lmsrCost`
//     numerically stable on imbalanced markets.
//
//   - The series (12 terms for exp, 14 for ln, in binary Q0.64 with constant
//     reciprocals) is the Rust's op for op, so results are bit-identical.
//     `tests/lmsr.test.ts` pins the small-buy case
//     `cost_delta(q=0, q=0, b=1000·WAD, d_yes=10·WAD, d_no=0)` to within
//     0.001% of ~5.0125 USDC, the same threshold as the Rust unit test
//     `math/lmsr.rs::cost_delta_golden_small_buy`.

export const WAD = 1_000_000_000_000_000_000n;
export const LN2_WAD = 693_147_180_559_945_309n;

// USDC = 6 decimals, WAD = 18 decimals; the ratio is the WAD↔USDC scalar.
// Architecture decision: reads return WAD; user-paid base-unit USDC is the
// `wadToUsdcCeil` ceiling at trade entry (matches `wad_to_usdc_ceil` in the
// on-chain `wad.rs`).
export const WAD_TO_USDC_SCALAR = 1_000_000_000_000n;

const EXP_MAX_INPUT_WAD = 64n * WAD;
// ⌊2⁶⁴ / n!⌋ for n = 2..12 and ⌊2⁶⁴ / (2n+1)⌋ for n = 1..13 — the same
// constants as `math/lmsr.rs`.
const INV_FACT_Q64 = [9223372036854775808n, 3074457345618258602n, 768614336404564650n, 153722867280912930n, 25620477880152155n, 3660068268593165n, 457508533574145n, 50834281508238n, 5083428150823n, 462129831893n, 38510819324n];
const INV_ODD_Q64 = [6148914691236517205n, 3689348814741910323n, 2635249153387078802n, 2049638230412172401n, 1676976733973595601n, 1418980313362273201n, 1229782938247303441n, 1085102592571150095n, 970881267037344821n, 878416384462359600n, 802032351030850070n, 737869762948382064n, 683212743470724133n];

export class LmsrMathError extends Error {
  constructor(msg: string) {
    super(`LmsrMathError: ${msg}`);
    this.name = "LmsrMathError";
  }
}

// ─── WAD primitives ────────────────────────────────────────────────────────

// Truncate-toward-zero division for signed bigints. `bigint / bigint` in JS
// already truncates toward zero (per the ECMAScript spec) so no shim needed.

export function wadMul(a: bigint, b: bigint): bigint {
  // (a * b) / WAD with rounding toward zero. JS bigint is arbitrary precision,
  // so the 256-bit intermediate machinery in the Rust port collapses to a
  // single multiply.
  return (a * b) / WAD;
}

export function wadDiv(a: bigint, b: bigint): bigint {
  if (b === 0n) throw new LmsrMathError("division by zero");
  // The program refuses divisors past 2^96 (its 256-bit division is exact
  // only below that); refuse them here too, so a port fails where it does.
  if ((b < 0n ? -b : b) > 1n << 96n) throw new LmsrMathError("wad_div: divisor out of range");
  return (a * WAD) / b;
}

// ─── exp ───────────────────────────────────────────────────────────────────

export function expWad(x: bigint): bigint {
  // Saturate the negative tail. exp(-EXP_MAX_INPUT_WAD) ≈ 1e-28; returning 0
  // keeps log-sum-exp stable for arbitrarily imbalanced markets.
  if (x < -EXP_MAX_INPUT_WAD) return 0n;
  if (x > EXP_MAX_INPUT_WAD) {
    throw new LmsrMathError(`exp_wad input out of domain: ${x}`);
  }
  if (x === 0n) return WAD;

  // Range-reduce: x = k · ln(2) + r, r ∈ [-ln(2)/2, ln(2)/2].
  let k: bigint;
  {
    const half = LN2_WAD / 2n;
    if (x >= 0n) {
      k = (x + half) / LN2_WAD;
    } else {
      k = (x - half) / LN2_WAD;
    }
  }
  const r = x - k * LN2_WAD;

  // exp(r) − 1 as a Taylor series in binary Q0.64 with constant reciprocal
  // factorials — no division in the loop, exactly as the Rust does it. Every
  // op here must round the way i128 does: `/` truncates toward zero and `>>`
  // floors, in both languages.
  const rq = (r << 64n) / WAD;
  let pow = rq;
  let acc = rq;
  for (const inv of INV_FACT_Q64) {
    pow = (pow * rq) >> 64n;
    acc += (pow * inv) >> 64n;
  }
  const sum = (((1n << 64n) + acc) * WAD) >> 64n;

  // Multiply by 2^k. k can be negative.
  if (k >= 0n) {
    const ku = Number(k);
    if (ku >= 127) throw new LmsrMathError(`exp_wad shift overflow: k=${k}`);
    return sum << BigInt(ku);
  } else {
    const ku = Number(-k);
    if (ku >= 127) return 0n;
    return sum >> BigInt(ku);
  }
}

// ─── ln ────────────────────────────────────────────────────────────────────

function bitLen(u: bigint): number {
  // Bit length of a non-negative bigint.
  if (u <= 0n) return 0;
  return u.toString(2).length;
}

export function lnWad(y: bigint): bigint {
  if (y <= 0n) throw new LmsrMathError(`ln_wad of non-positive: ${y}`);
  // y = 2^k · m, m ∈ [WAD, 2·WAD).
  const yu = y;
  const bits = bitLen(yu);
  // WAD bit length is 60.
  const wadBits = bitLen(WAD);
  const k = bits - wadBits;
  let m: bigint;
  if (k >= 0) {
    m = yu >> BigInt(k);
  } else {
    const kk = -k;
    if (kk >= 127) throw new LmsrMathError(`ln_wad shift overflow`);
    m = yu << BigInt(kk);
  }
  // ln m = 2·atanh(z), z = (m − 1)/(m + 1), in Q0.64 with constant
  // reciprocal odds — the Rust's series, op for op.
  const mMinus = m - WAD;
  const mPlus = m + WAD;
  const z = (mMinus << 64n) / mPlus;
  const z2 = (z * z) >> 64n;
  let term = z;
  let acc = z;
  for (const inv of INV_ODD_Q64) {
    term = (term * z2) >> 64n;
    acc += (term * inv) >> 64n;
  }
  const lnMOverWad = ((2n * acc) * WAD) >> 64n;
  const kTerm = BigInt(k) * LN2_WAD;
  return kTerm + lnMOverWad;
}

// ─── Cost ──────────────────────────────────────────────────────────────────

// `C(q_yes, q_no, b) = b · ln(exp(q_yes/b) + exp(q_no/b))` with the
// log-sum-exp shift baked in. Inputs/output are WAD-scaled.
export function lmsrCost(qYes: bigint, qNo: bigint, b: bigint): bigint {
  if (b <= 0n) throw new LmsrMathError(`lmsr_cost b must be > 0, got ${b}`);
  const qyOverB = wadDiv(qYes, b);
  const qnOverB = wadDiv(qNo, b);
  const m = qyOverB > qnOverB ? qyOverB : qnOverB;
  const eYes = expWad(qyOverB - m);
  const eNo = expWad(qnOverB - m);
  const sum = eYes + eNo;
  const lnSum = lnWad(sum);
  const inner = m + lnSum;
  return wadMul(b, inner);
}

// `C(q + Δ) - C(q)`. Positive return = user pays this much. Negative = sell
// proceeds.
export function costDelta(
  qYes: bigint,
  qNo: bigint,
  b: bigint,
  dYes: bigint,
  dNo: bigint,
): bigint {
  const before = lmsrCost(qYes, qNo, b);
  const after = lmsrCost(qYes + dYes, qNo + dNo, b);
  return after - before;
}

// Convert a non-negative WAD amount to USDC base units, rounding **up**.
// Mirrors `wad_to_usdc_ceil` in `sooth_core`'s `math/wad.rs`.
export function wadToUsdcCeil(wad: bigint): bigint {
  if (wad < 0n) throw new LmsrMathError(`wad_to_usdc_ceil negative: ${wad}`);
  const s = WAD_TO_USDC_SCALAR;
  return (wad + s - 1n) / s;
}

// Floor variant. Redemption displays round DOWN so a shown payout is never
// larger than the one the vault will actually pay.
export function wadToUsdcFloor(wad: bigint): bigint {
  if (wad < 0n) throw new LmsrMathError(`wad_to_usdc_floor negative: ${wad}`);
  return wad / WAD_TO_USDC_SCALAR;
}

// YES probability after a hypothetical (qYes, qNo, b) state. WAD-scaled.
// p_yes = exp(q_yes/b) / (exp(q_yes/b) + exp(q_no/b)). With log-sum-exp
// shift, p_yes = exp(qy/b - m) / (exp(qy/b - m) + exp(qn/b - m)).
export function yesPriceWad(qYes: bigint, qNo: bigint, b: bigint): bigint {
  if (b <= 0n) throw new LmsrMathError(`yes_price_wad b must be > 0`);
  const qyOverB = wadDiv(qYes, b);
  const qnOverB = wadDiv(qNo, b);
  const m = qyOverB > qnOverB ? qyOverB : qnOverB;
  const eYes = expWad(qyOverB - m);
  const eNo = expWad(qnOverB - m);
  const sum = eYes + eNo;
  if (sum === 0n) return WAD / 2n;
  return wadDiv(eYes, sum);
}

File: packages/programs-core/programs/sooth-core/src/math/lmsr.rs
//! LMSR cost function — `exp_wad`, `ln_wad`, and the log-sum-exp shifted
//! `cost_delta` used by `trade_positions`.
//!
//! The math model is the standard binary-outcome LMSR with `b` (liquidity
//! parameter) in WAD:
//!
//! ```text
//! C(q_yes, q_no, b) = b · ln( exp(q_yes/b) + exp(q_no/b) )
//! ```
//!
//! ## Numerical stability
//!
//! The "tail 100× cheaper than 10×" property comes from the **log-sum-exp
//! shift**: subtract `m = max(q_yes/b,
//! q_no/b)` before exp, add it back outside ln. Both exp arguments end up ≤
//! 0; for very imbalanced markets the smaller side falls below the
//! `EXP_MAX_INPUT_WAD` saturation clamp and short-circuits to 0 without
//! running the Taylor series. That is what makes the 100× tail (33k CU)
//! cheaper than the 10× imbalanced case (55k CU).

pub use super::wad::MathError;
use super::wad::{wad_div, wad_mul, LN2_WAD, WAD, WAD_U};

/// Maximum |x| for `exp_wad`. ~133·WAD is where exp(x) approaches u128::MAX;
/// we clamp tighter at 64·WAD because LMSR diffs past that mean a one-sided
/// market that's essentially resolved. exp(-64·WAD) ≈ 1e-28 is below any
/// WAD-representable value, so `exp_wad` saturates the negative tail to 0.
pub const EXP_MAX_INPUT_WAD: i128 = 64 * WAD;

/// Number of Taylor terms for `exp`. 12 gives ~1e-18 relative error on the
/// reduced range |r| ≤ ln(2)/2 ≈ 0.347·WAD.
const EXP_TERMS: u32 = 12;

/// Number of Taylor terms for `ln` (in `(z + z³/3 + z⁵/5 + …)` form).
/// 14 gives ~1e-19 relative error on z ∈ [0, 1/3).
const LN_TERMS: u32 = 14;

pub fn exp_wad(x: i128) -> Result<i128, MathError> {
    // Saturate the negative tail. exp(-EXP_MAX_INPUT_WAD) ≈ 1e-28; returning
    // 0 keeps log-sum-exp stable for arbitrarily imbalanced markets and is
    // what lets the "tail" CU case short-circuit (see module comment).
    if x < -EXP_MAX_INPUT_WAD {
        return Ok(0);
    }
    if x > EXP_MAX_INPUT_WAD {
        return Err(MathError::Overflow);
    }
    if x == 0 {
        return Ok(WAD);
    }
    // Range reduction: x = k * ln(2) + r, r ∈ [-ln(2)/2, ln(2)/2].
    let k = {
        let half = LN2_WAD / 2;
        if x >= 0 {
            (x + half) / LN2_WAD
        } else {
            (x - half) / LN2_WAD
        }
    };
    let r = x - k * LN2_WAD;

    // Taylor: exp(r) = Σ rⁿ / n!
    let mut term = WAD;
    let mut sum = WAD;
    for n in 1..=EXP_TERMS {
        term = wad_mul(term, r)?;
        term /= n as i128;
        sum = sum.checked_add(term).ok_or(MathError::Overflow)?;
    }
    // Multiply by 2^k.
    let result = if k >= 0 {
        let k_u = k as u32;
        if k_u >= 127 {
            return Err(MathError::Overflow);
        }
        sum.checked_shl(k_u).ok_or(MathError::Overflow)?
    } else {
        let k_u = (-k) as u32;
        if k_u >= 127 {
            return Ok(0);
        }
        sum >> k_u
    };
    Ok(result)
}

pub fn ln_wad(y: i128) -> Result<i128, MathError> {
    if y <= 0 {
        return Err(MathError::Overflow);
    }
    let yu = y as u128;
    let bits = 128 - yu.leading_zeros() as i32;
    let wad_bits = 128 - WAD_U.leading_zeros() as i32; // 60
    let k = bits - wad_bits;
    let m: i128 = if k >= 0 {
        (yu >> (k as u32)) as i128
    } else {
        let kk = (-k) as u32;
        if kk >= 127 {
            return Err(MathError::Overflow);
        }
        (yu << kk) as i128
    };
    let m_minus = m - WAD;
    let m_plus = m + WAD;
    let z = wad_div(m_minus, m_plus)?;
    let z2 = wad_mul(z, z)?;
    let mut term = z;
    let mut sum = z;
    for n in 1..LN_TERMS {
        term = wad_mul(term, z2)?;
        let denom = (2 * n + 1) as i128;
        sum = sum.checked_add(term / denom).ok_or(MathError::Overflow)?;
    }
    let ln_m_over_wad = sum.checked_mul(2).ok_or(MathError::Overflow)?;
    let k_term = (k as i128)
        .checked_mul(LN2_WAD)
        .ok_or(MathError::Overflow)?;
    Ok(k_term + ln_m_over_wad)
}

/// `C(q_yes, q_no, b) = b · ln( exp(q_yes/b) + exp(q_no/b) )` with the
/// log-sum-exp shift baked in. Inputs and output WAD-scaled.
#[inline(always)]
pub fn lmsr_cost(q_yes: i128, q_no: i128, b: i128) -> Result<i128, MathError> {
    let qy_over_b = wad_div(q_yes, b)?;
    let qn_over_b = wad_div(q_no, b)?;
    let m = qy_over_b.max(qn_over_b);
    let e_yes = exp_wad(qy_over_b - m)?;
    let e_no = exp_wad(qn_over_b - m)?;
    let sum = e_yes.checked_add(e_no).ok_or(MathError::Overflow)?;
    let ln_sum = ln_wad(sum)?;
    let inner = m.checked_add(ln_sum).ok_or(MathError::Overflow)?;
    wad_mul(b, inner)
}

/// Signed cost of moving the AMM state from `(q_yes, q_no)` to
/// `(q_yes + d_yes, q_no + d_no)`.
///
/// Positive return = user pays this much. Negative = sell proceeds.
#[inline(always)]
pub fn cost_delta(
    q_yes: i128,
    q_no: i128,
    b: i128,
    d_yes: i128,
    d_no: i128,
) -> Result<i128, MathError> {
    let before = lmsr_cost(q_yes, q_no, b)?;
    let q_yes_after = q_yes.checked_add(d_yes).ok_or(MathError::Overflow)?;
    let q_no_after = q_no.checked_add(d_no).ok_or(MathError::Overflow)?;
    let after = lmsr_cost(q_yes_after, q_no_after, b)?;
    Ok(after - before)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: i128, b: i128, eps: i128) -> bool {
        (a - b).abs() <= eps
    }

    #[test]
    fn exp_zero_is_one() {
        assert_eq!(exp_wad(0).unwrap(), WAD);
    }

    #[test]
    fn exp_one_is_e() {
        let e = exp_wad(WAD).unwrap();
        // e ≈ 2.718281828... × WAD
        assert!(close(e, 2_718_281_828_459_045_235, 1_000_000));
    }

    #[test]
    fn exp_negative_one() {
        let v = exp_wad(-WAD).unwrap();
        // 1/e ≈ 0.367879441 × WAD
        assert!(close(v, 367_879_441_171_442_321, 1_000_000));
    }

    #[test]
    fn exp_saturates_negative_tail() {
        // Past the saturation clamp we return 0 — keeps the log-sum-exp shift
        // numerically stable for 100×-imbalanced markets.
        assert_eq!(exp_wad(-(EXP_MAX_INPUT_WAD + 1)).unwrap(), 0);
    }

    #[test]
    fn ln_one_is_zero() {
        assert_eq!(ln_wad(WAD).unwrap(), 0);
    }

    #[test]
    fn ln_e_is_one() {
        let e_wad = 2_718_281_828_459_045_235;
        let v = ln_wad(e_wad).unwrap();
        assert!(close(v, WAD, 1_000_000));
    }

    #[test]
    fn ln_two_is_ln2() {
        let v = ln_wad(2 * WAD).unwrap();
        assert!(close(v, LN2_WAD, 1_000_000));
    }

    #[test]
    fn round_trip_ln_exp() {
        for &x in &[WAD / 4, WAD / 2, WAD, 2 * WAD, 5 * WAD] {
            let e = exp_wad(x).unwrap();
            let back = ln_wad(e).unwrap();
            assert!(
                close(back, x, 10_000_000),
                "ln(exp({x})) = {back}, expected {x}"
            );
        }
    }

    #[test]
    fn lmsr_cost_balanced_book() {
        // Symmetric q: C(q,q,b) = b · ln(2 · exp(q/b · -m + m)) shifted; the
        // closed form is C(q,q,b) = q + b·ln(2). Verify within 1e-9 WAD.
        let b = 1_000 * WAD;
        let q = 500 * WAD;
        let c = lmsr_cost(q, q, b).unwrap();
        // expected = q + b · ln 2
        let expected = q + (b / WAD) * LN2_WAD; // b in WAD, so multiply by ln2 then back to WAD
                                                // The cleaner expected: b·ln 2 in WAD = wad_mul(b, LN2_WAD)
        let expected_clean = q + wad_mul(b, LN2_WAD).unwrap();
        let _ = expected;
        assert!(
            close(c, expected_clean, 1_000_000_000),
            "got {c} vs {expected_clean}"
        );
    }

    #[test]
    fn cost_delta_buy_yes_increases_cost() {
        // Buying YES from a balanced book costs > 0 and < δ (since price < 1).
        let b = 1_000 * WAD;
        let q = 500 * WAD;
        let delta = b / 100; // 1% of b
        let dc = cost_delta(q, q, b, delta, 0).unwrap();
        assert!(dc > 0);
        assert!(dc < delta);
    }

    /// Golden-value pin. `cost_delta(q=0, q=0, b=1000·WAD,
    /// d_yes=10·WAD, d_no=0)` — the small-buy case from the bench grid.
    #[test]
    fn cost_delta_golden_small_buy() {
        let b = 1_000 * WAD;
        let dc = cost_delta(0, 0, b, b / 100, 0).unwrap();
        // From a balanced empty book, price is 0.5; buying 10 shares of YES
        // costs ≈ 5 + (δ²/(2·b·ln2)) · ln2 ≈ 5 + 0.0125 = 5.0125 USDC. The
        // exact LMSR value lands at 5_012_499_947_917_016_000 WAD on this
        // toolchain (Taylor truncation residue ≈ 5e-12). Loose range so
        // Taylor drift across compiler versions doesn't break the test.
        assert!(
            dc > 5_010_000_000_000_000_000 && dc < 5_015_000_000_000_000_000,
            "cost_delta out of range: {dc}"
        );
    }
}

#[cfg(test)]
mod wide_cost_regression {
    use super::*;

    /// Buying shares must never produce a negative cost.
    ///
    /// `lmsr_cost` ends in `wad_mul(b, m + ln_sum)`, and that product exceeds
    /// `u128::MAX` once the cost passes ~340.28e18 — which is just
    /// `max(q) + b·ln(2)`, i.e. a market with a few hundred shares
    /// outstanding. `wad_mul` must error there rather than wrap.
    ///
    /// `cost_delta` differences two costs, so equal wraps at both endpoints
    /// cancel exactly and a market looks healthy. A trade straddling the
    /// boundary does not cancel: at b = 50, q = (307, 303), buying 4 YES
    /// yields **-338.16** — the program paying the trader to take shares.
    #[test]
    fn buying_never_costs_a_negative_amount() {
        for b in [1i128, 50, 100, 1_000, 10_000] {
            let b = b * WAD;
            for q in [0i128, 100, 300, 320, 340, 400, 1_000, 5_000] {
                for d in [1i128, 4, 50, 500] {
                    let c = cost_delta(q * WAD, q * WAD, b, d * WAD, 0)
                        .unwrap_or_else(|e| panic!("b={b} q={q} d={d}: {e:?}"));
                    assert!(c > 0, "b={b} q={q} d={d} gave cost {c}");
                }
            }
        }
    }

    /// Cost is bounded by the number of shares bought.
    ///
    /// A share can never cost more than 1 unit under LMSR, so this catches a
    /// wrap in the other direction — one that inflates the charge rather than
    /// inverting it.
    #[test]
    fn buying_never_costs_more_than_one_per_share() {
        for b in [1i128, 50, 1_000] {
            let b = b * WAD;
            for q in [0i128, 300, 340, 1_000] {
                for d in [1i128, 4, 50, 500] {
                    let c = cost_delta(q * WAD, q * WAD, b, d * WAD, 0).unwrap();
                    assert!(c <= d * WAD, "b={b} q={q} d={d} gave cost {c}");
                }
            }
        }
    }

    /// A balanced market's cost has a closed form: `q + b·ln(2)`.
    ///
    /// Checked straight through the wrap boundary, since that is the only
    /// place the limb arithmetic can disagree with the identity.
    #[test]
    fn balanced_cost_matches_the_closed_form() {
        let b = 50 * WAD;
        for q in [0i128, 100, 300, 320, 339, 340, 341, 400, 1_000] {
            let got = lmsr_cost(q * WAD, q * WAD, b).unwrap();
            let want = q * WAD + wad_mul(b, LN2_WAD).unwrap();
            let diff = (got - want).abs();
            assert!(diff < 1_000_000, "q={q}: got {got}, want {want}");
        }
    }

    /// Buying then selling the same size returns to the same cost.
    ///
    /// A wrap is not symmetric, so a round trip across the boundary would lose
    /// (or create) value.
    #[test]
    fn a_round_trip_is_value_neutral() {
        let b = 50 * WAD;
        for q in [300i128, 320, 340] {
            let buy = cost_delta(q * WAD, q * WAD, b, 10 * WAD, 0).unwrap();
            let sell = cost_delta((q + 10) * WAD, q * WAD, b, -10 * WAD, 0).unwrap();
            assert_eq!(buy, -sell, "q={q}");
        }
    }
}
